subsys/usb/usb_c/usbc_pe_common.c

```diff
@@ -229,6 +229,39 @@ void pe_message_sent(const struct device *dev)
 	atomic_set_bit(pe->flags, PE_FLAGS_TX_COMPLETE);
 }
 
+static bool pe_soft_reset_is_required(const struct device *dev, const enum pd_packet_type type)
+{
+	struct usbc_port_data *data = dev->data;
+	struct policy_engine *pe = data->pe;
+
+	/* Protocol Error not on SOP */
+	if (type != PD_PACKET_SOP) {
+		return false;
+	}
+
+	if (atomic_test_bit(pe->flags, PE_FLAGS_EXPLICIT_CONTRACT)) {
+		/*
+		 * If the first Message in an AMS has been passed to the
+		 * Protocol Layer by the Policy Engine but has not yet been sent
+		 * when the Protocol Error occurs, the Policy Engine Shall Not
+		 * issue a Soft Reset
+		 */
+		if (!atomic_test_bit(pe->flags, PE_FLAGS_FIRST_MSG_SENT)) {
+			return false;
+		}
+
+		/*
+		 * If the Protocol Error occurs during an Interruptible AMS then
+		 * the Policy Engine Shall Not issue a Soft Reset
+		 */
+		if (atomic_test_bit(pe->flags, PE_FLAGS_INTERRUPTIBLE_AMS)) {
+			return false;
+		}
+	}
```

Comment:
It would be better to follow the language of the specification, even if this means the code is little bit larger.

```
if (SOP && !INTERRUPTIBLE_AMS)
    return true;

/* TODO - Check if a message has not been sent after retries to the Sink */

if (!EXPLICIT_CONTRACT) {
    if (SOP && INTERRUPTIBLE_AMS)
        return true;
    if (SOP && !FIRST_MESSAGE_SENT)
        return true; 
}
return false;
```

Reply:
From our offline converstion, the spec describes both implementations.